Add support for listing and retrieving images

samuelallan72 opened the pull request:
note: depends on https://github.com/canonical/nlxd-rs/pull/15

Commit:
Add support for listing and retrieving images

TheSignPainter98 commented on examples/images.rs:
```diff
@@ -0,0 +1,12 @@
+fn main() {
+    let client = Client::new(Default::default()).unwrap();
+    let image_fingerprints: Vec<String> = client.images().unwrap();
+    println!("{:#?}", &image_fingerprints);
```
Outputting with `Debug` in an example looks like something's slightly off. Also these two `println!` calls seem to output more or less the same thing? Perhaps adding some text into the print statement and outputting something specific could be useful?

TheSignPainter98 commented on src/client.rs:
```diff
@@ -202,4 +203,24 @@ impl Client {
     }
+
+    /// Get a list of image IDs
+    pub fn images(&self) -> Result<Vec<String>> {
```
I think we can help out the person using this API by replacing `String` here with some `ImageID` type. Also, it seems problematic that `client.images()` doesn't actually return images... this is actually nicely demonstrated by the example, where the return of this function has to be assigned into `image_fingerprints` rather than some `images` declaration ;)

TheSignPainter98 replied:
We could instead return some lazy `Image` type, where accessing its properties would be done by fallible method. This could work nicely, I'm just concerned about timing---the first time we access a property, we'd effectively take a snapshot of the image, so future property accesses wouldn't get updated. This could be resolved with some `refresh` method to force the cached value to be updated, but I'm wondering would this be too arcane?

What if instead, `images` returned an `impl Iterator<Item = Result<Image>>`? So it's obvious that when the item is extracted from the iterator that's it at that time? If the user didn't want to get the images, there could be a separate `fn image_ids(&self) -> impl Iterator<Item = ImageId>`, where we'd have some way to get an image using a given image ID (perhaps some `Image::retreive(image_id)`)?

samuelallan72 replied:
Thanks for these notes.  I don't think I'm a fan of the first option (lazy image type) due to the concerns about caching and refreshing you noted.  I'd also prefer any network requests be explicit, and a network request on accessing an attribute seems unexpected.

I like the idea of returning an iterator, that could be interesting.  We can definitely have a separate image_ids method.

> perhaps some `Image::retreive(image_id)`

Ah interesting, this goes back to a design discussion that I'm not sure we resolved completely?  To pattern things around a single client with methods vs 'objects' with methods to retrieve related things.  eg.

- `client.get_image_by_id(image_id)`
- vs `Image::retrieve(client_config, image_id)`

TheSignPainter98 commented on src/client.rs:
```diff
@@ -202,4 +203,24 @@ impl Client {
+            .get::<Vec<String>>(&format!("{}/images", self.version.to_url_segment()))?
+            .metadata
+            .into_iter()
+            .filter_map(|s| s.strip_prefix("/1.0/images/").map(|s| s.to_owned()))
```
Can the `.map(|s| ...)` be moved out of the `.filter_map`? I think flatter call chains are easier to read as it encourages simple steps.

Also, what happens if an images _doesn't_ have the `/1.0/images/` prefix? Is it worth returning an error in this case?

TheSignPainter98 commented on src/images.rs:
```diff
@@ -0,0 +1,111 @@
+// NOTE: this can be moved to a shared types file later.
+/// https://github.com/canonical/lxd/blob/main/doc/architectures.md
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
```
I think `snake_case` relies on their being no uppercase letters after the start of these enum variants, perhaps `lowercase` would be cleaner?

TheSignPainter98 commented on src/images.rs:
```diff
@@ -0,0 +1,111 @@
+// NOTE: this too can be moved to a shared location
+#[allow(unused)]
+#[derive(Debug, Deserialize)]
+pub struct Bytes(u64);
```
The name `Bytes` represents the what but doesn't hint at the why, when reading this, I would have expected the content to have been `Vec<u8>` rather than `u64`. Instead, perhaps we could call this `SizeBytes`? I'm actually unsure whether we need this extra type, the doc string on the `size` field already covers this. Perhaps that field could be renamed to `size_bytes`, `num_bytes` or `nbytes`?

TheSignPainter98 commented on src/images.rs:
```diff
@@ -0,0 +1,111 @@
+    auto_update: bool,
+    /// Whether the image is an automatically cached remote image
+    cached: bool,
+    // TODO: these should be datetimes - added in https://github.com/NucciTheBoss/nlxd/pull/19/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542
```
This points to a closed issue :) I think it would be better to add a comment to this PR to the effect of 'This PR awaits #15', then fix this before merging. I think this is a good case to put this PR back into draft just to avoid any accidental merging

TheSignPainter98 commented on src/images.rs:
```diff
@@ -0,0 +1,111 @@
+pub struct ImageSource {
+    alias: String,
+    certificate: String,
+    /// NOTE: lxd api docs say:
+    /// Type of image (container or virtual-machine).
+    /// However this field has observed to also be an empty string.
+    image_type: String,
```
That sounds like this field should have type `Option<String>` :P Is it possible to change this? If so, I think this comment is redundant.

Hold on, we have the `ImageType` type here, could this be used here?

samuelallan72 replied:
True, this can absolutely be an `Option<ImageType>`.  I don't know what it means for the image type to be empty here though, so I opened an issue on lxd: [canonical/lxd#13586](https://github.com/canonical/lxd/issues/13586).

TheSignPainter98 commented on src/images.rs:
```diff
@@ -0,0 +1,111 @@
+#[serde(rename_all = "snake_case")]
+pub enum RemoteImageProtocol {
+    Lxd,
+    Simplestreams,
```
The name `Simplestreams` looks awkward, would `rename_all = "lowercase"` fix this? It would be nice to capitalise this on the Rust side

TheSignPainter98 reviewed (changes requested):
This is coming along nicely! Here are a few points to think about

Commit:
Use serde lowercase rename

Lowercase is the intention,
and this avoids issues if the rust type has an uppercase somewhere else.

Commit:
Use proper type for image_type

image_type can sometimes be an empty string though,
so map this to `None`.
I'm unsure if an empty string is valid,
or what it means,
so this is tracked at https://github.com/canonical/lxd/issues/13586

samuelallan72 commented:
@TheSignPainter98 @NucciTheBoss I'll actually close my PRs here now that I've left the company and no longer have access to this repository. If I contribute in the future, I can open PRs from a personal fork. :)
